R/valley.R

```diff
@@ -0,0 +1,299 @@
+}
+
+#'Write dem to cloud optimized GeoTiff file as specified location
+#'
+#' @param dem to write to file
+#' @param fpath filepath for output. If no output directory is specified
+#' (see below) fpath is parsed to determine
+#' the output directory
+#' @param output_directory where file should be written.
+#' If specified fpath is treated as filename only.
+#'
+#' @export
+dem_to_cog <- function(dem, fpath, output_directory = NULL) {
+  if (is.null(output_directory)) {
+    filename <- basename(fpath)
+    directory_name <- dirname(file)
+  } else {
+    filename <- fpath
+    directory_name <- output_directory
+  }
+  data_dir  <- here::here(directory_name)
+  terra::writeRaster(
+                     x = dem,
+                     filename = sprintf("%s/%s", data_dir, filename),
+                     filetype = "COG",
+                     overwrite = TRUE)
+}
+
+
+
+
+
+
+#' Reproject a raster or vector dataset to the specified
+#' coordinate reference system (CRS)
+#'
+#' @param x Raster or vector object
+#' @param crs CRS to be projected to
+#' @param ... Optional arguments for raster or vector reproject functions
+#'
+#' @return Object reprojected to specified CRS
+#' @export
+reproject <- function(x, crs, ...) {
+  if (inherits(x, "SpatRaster")) {
+    return(terra::project(x, crs, ...))
+  } else if (inherits(x, c("bbox", "sfc", "sf"))) {
+    return(sf::st_transform(x, crs, ...))
+  } else {
+    stop(sprintf("Cannot reproject object type: %s", class(x)))
+  }
+}
+
+#' spatially smooth dem by (window) filtering
+#'
+#' @param dem raster data of dem
+#' @param method smoothing function to be used, e.g. "median".
+#' As accepted by focal
+#' @param window size of smoothing kernel
+#'
+#' @return filtered dem
+#' @export
+smooth_dem <- function(dem, method = "median", window = 5) {
+  dem_smoothed <- terra::focal(dem, w = window, fun = method)
+  names(dem_smoothed) <- "dem_smoothed"
+  return(dem_smoothed)
+}
+
+#' Derive slope as percentage from dem
+#' This makes use of the terrain function of the terra package
+#'
+#' @param dem raster data of dem
+#'
+#' @return raster of derived slope over dem extent
+#' @export
+get_slope <- function(dem) {
+  slope_radians <- terra::terrain(dem, v = "slope", unit = "radians")
+  slope <- tan(slope_radians)
+  return(slope)
+}
+
+#' Mask slope raster, setting the slope to zero for the pixels overlapping
+#' the river area.
+#'
+#' @param slope raster data of slope
+#' @param vector/polygon data of river
+#' @param lthresh lower numerival threshold to consider slope non-zero
+#'
+#' @return updated slope raster
+#' @export
+mask_slope <- function(slope, river, lthresh = 1.e-3, target = 0) {
+  slope_masked <- terra::mask(
+                              slope,
+                              terra::ifel(slope <= lthresh, NA, 1),
+                              updatevalue = lthresh)
+
+  slope_masked <- terra::mask(
+                              slope_masked,
+                              river,
```

Comment:
```suggestion
                              terra::vect(river),
```
This is needed so that `slope_masked ` and `river` are compatible in `terra::mask()`